### PR TITLE
fix: issue D401 only for non-test/property functions and methods

### PR DESCRIPTION
**Pull request by scop**
Extend test fixture to verify the targeting.

Includes two "attribute docstrings" which per PEP 257 are not recognized by the Python bytecode compiler or available as runtime object attributes. They are not available for us either at time of writing, but include them for completeness anyway in case they one day are.

Rust newbie here, review with a fine comb greatly appreciated :)

### CODE REVIEW

File: src/rules/pydocstyle/rules/non_imperative_mood.rs [hunk 1]
```diff
@@ -14,6 +15,38 @@ static MOOD: Lazy<Mood> = Lazy::new(Mood::new);
...
+            decorator_list,
+            ..
+        } => {
+            if name.starts_with("test") || name.eq("runTest") {
+                return;
+            }
+            for expr in decorator_list.iter() {
+                if let ExprKind::Name { id, .. } = &expr.node {
+                    if id.eq("property") {
+                        return;
```
Comment (charliermarsh) [inline src/rules/pydocstyle/rules/non_imperative_mood.rs:42]: It looks like Pydocstyle defaults to accepting `property,cached_property,functools.cached_property`.

Maybe just for consistency with this "kind" of check, let's add a method to `src/visibilty.rs` like:

```rs
pub fn is_property(checker: &Checker, decorator_list: &[Expr]) -> bool {
    decorator_list.iter().any(|expr| {
        checker.resolve_call_path(expr).map_or(false, |call_path| {
            call_path.as_slice() == ["", "property"] || call_path.as_slice() == ["functools", "cached_property"]
        })
    })
}
```

That will check the additional decorator _and_ ensure that it only triggers when `functools` is imported _and_ track aliases _and_ track overrides :)

**Review by charliermarsh — CHANGES_REQUESTED** [review]
Great! Let me know if any of the below is unclear.

File: src/rules/pydocstyle/rules/non_imperative_mood.rs (hunk 1 above)
Comment (scop) [reply src/rules/pydocstyle/rules/non_imperative_mood.rs:42]: I actually implemented `is_property` in `visibility.rs` in my first cut of this patch, but then decided to move it back, "inlined" at the call site. I felt `@property` isn't actually related to visibility in the broad sense the other similar helpers there are, and thus it seemed misplaced.

But no strong opinions really, could you confirm if you'd still rather see it there and I'll refactor.

In any case, will revise this tomorrow, addressing the other stuff. Thanks!

Comment (charliermarsh) [reply src/rules/pydocstyle/rules/non_imperative_mood.rs:42]: Yeah I think that's a totally reasonable instinct, but I'd still rather see it in there. There are some other helpers in that file that aren't _really_ related to visibility (like `is_override`), and I'd like to have them logically grouped for now.

File: src/rules/pydocstyle/rules/non_imperative_mood.rs [hunk 2]
```diff
@@ -14,6 +15,38 @@ static MOOD: Lazy<Mood> = Lazy::new(Mood::new);
...
+            name,
+            decorator_list,
+            ..
+        }
+        | StmtKind::AsyncFunctionDef {
+            name,
+            decorator_list,
+            ..
+        } => {
+            if name.starts_with("test") || name.eq("runTest") {
```
[the remark this answers is not in the document]
Comment (scop) [reply src/rules/pydocstyle/rules/non_imperative_mood.rs]: Made use of that and `cast::decorator_list` in 05ec78c0812d8657442a6cd4900dfd235da17b50.

File: src/rules/pydocstyle/rules/non_imperative_mood.rs (hunk 1 above)
Comment (scop) [reply src/rules/pydocstyle/rules/non_imperative_mood.rs]: Sure, done in f38b51a29dc243060f77c6d90e6e453242f1ef7e.

**Comment by charliermarsh** [comment]
Looks great! Will merge when CI is green.

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
Thanks tons.

**Comment by akx** [comment]
@scop Thanks for refining this! ❤️

**Pull request merged**